Add impl Encode for [T], where T: Encode

cronokirby opened the pull request:
Since Encode takes a reference, this allows us to encode &[T] directly
using this implementation. The encoding scheme is the same as for
Vec<T>.

This also makes the implementation for &[u8] superfluous, since we get
an implementation for [u8] by virtue of u8 implementing encode. This
also gives us free implementations for &[u16], &[u32], etc. which is
quite useful. Nonetheless, we keep the implementation for &[u8] around,
because the implementation can directly write a large number of bytes,
which can be more efficient than the generic implementation.

And uh, apologies if you would have preferred I raise an issue first, but this felt like a small enough change to warrant a direct PR.

I actually ran into a situation where this was convenient. My workaround to not having this was to take `&Vec<T>` as an argument to my function, and this seemed highly unidiomatic. I was also surprised not to see this when digging around in the docs the other day.

cronokirby committed:
Add impl Encode for [T], where T: Encode

Since Encode takes a reference, this allows us to encode &[T] directly
using this implementation. The encoding scheme is the same as for
Vec<T>.

This also makes the implementation for &[u8] superfluous, since we get
an implementation for [u8] by virtue of u8 implementing encode. This
also gives us free implementations for &[u16], &[u32], etc. which is
quite useful. Nonetheless, we keep the implementation for &[u8] around,
because the implementation can directly write a large number of bytes,
which can be more efficient than the generic implementation.

VictorKoenders commented:
I think this will conflict with #526 and I want to get that PR merged in first, so this PR will have to stay open for a while.

Thank you for your contribution regardless!

VictorKoenders commented:
#526 is now merged.

Looks like this PR is still valid. You can remove the implementation on `Box<[T]>` because I think the implementation on `[T]` and `Box<T>` is strictly better.

cronokirby committed:
Remove redundant Encode implementations

Since we've implemented Encode for [T], this makes the implementation
for Box<[T]> redundant (since we have a blanket implementation for
Box<T>), and ditto for &[T], which this change replaces by combining the
implementations for [T] and &T.

cronokirby commented:
Nice! I've removed redundant implementations for `Box<T>` as well as for `&[T]` (since it's made redundant since we now have both `[T]` and `&T`, and thus `&[T]`), the behavior for the latter is the same as before, so this should have no effect on end users.

VictorKoenders commented on src/enc/impls.rs:
```diff
@@ -324,26 +337,6 @@ fn encode_utf8(writer: &mut impl Writer, c: char) -> Result<(), EncodeError> {
     }
 }
 
-// BlockedTODO: https://github.com/rust-lang/rust/issues/37653
-//
-// We'll want to implement encoding for both &[u8] and &[T: Encode],
-// but those implementations overlap because u8 also implements Encode
-// impl Encode for &'_ [u8] {
-//     fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
-//         encoder.writer().write(*self)
-//     }
-// }
```
This comment can stay, we still want to do implementation specialization

cronokirby replied:
Maybe move the comment above the [T] implementation then?

cronokirby committed:
Reinclude comment about Encode specialization for &[u8]

VictorKoenders commented:
Thanks!
